access_amherst_backend/access_amherst_algo/email_scraper/email_parser.py

```diff
@@ -0,0 +1,156 @@
+        res, msg = mail.fetch(msg_num, "(RFC822)")
+        for response_part in msg:
+            if isinstance(response_part, tuple):
+                msg = email.message_from_bytes(response_part[1])
+                return msg
+    return None
+
+# Function to extract the email body
+def extract_email_body(msg):
+    if msg.is_multipart():
+        for part in msg.walk():
+            content_type = part.get_content_type()
+            if content_type == "text/plain":
+                return part.get_payload(decode=True).decode('utf-8')
+    else:
+        return msg.get_payload(decode=True).decode('utf-8')
+
+# Function to send email content to Gemini API and get JSON response
+def extract_event_info_using_gemini(email_content):
+    prompt = f"""
+        {email_content}
+    """
+    
+    # Use Gemini API to generate the response
+    response = model.generate_content(prompt)
+
+    return response.text
+
+def save_to_json_file(data, filename, folder):
+    # Ensure the folder exists
+    if not os.path.exists(folder):
+        os.makedirs(folder)
+    
+    # Construct the full file path
+    file_path = os.path.join(folder, filename)
+
+    try:
+        with open(file_path, 'w') as json_file:
+            json.dump(data, json_file, indent=4)  # Save data with indentation for readability
+        print(f"Data successfully saved to {file_path}")
+    except Exception as e:
+        print(f"Failed to save data to {file_path}: {e}")
+
+def parse_email(subject_filter):
+    app_password = os.getenv("EMAIL_PASSWORD")
+    subject_filter = subject_filter
+
+    msg = connect_and_fetch_latest_email(app_password, subject_filter)
+    if msg:
+        email_body = extract_email_body(msg)
+        print("Email fetched successfully.")
+        
+        extracted_events_json = extract_event_info_using_gemini(email_body)
+        
+        if extracted_events_json:  # Ensure data exists
+            try:
+                # Convert the string response to a Python object
+                events_data = json.loads(extracted_events_json)
+
+                # Generate a timestamped filename
+                timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
+                filename = f"extracted_events_{timestamp}.json"
+
+                # get current dir
+                curr_dir = os.path.dirname(os.path.abspath(__file__))
+                
+                # Define the relative path to the json_outputs directory
+                output_dir = os.path.join(curr_dir, "json_outputs")
+                
+                # Ensure the directory exists
+                if not os.path.exists(output_dir):
+                    os.makedirs(output_dir)
+
+                # Create the full file path
+                file_path = os.path.join(output_dir, filename)
+                
+                # Save the extracted events to a JSON file
+                save_to_json_file(events_data, filename, output_dir)
+            except json.JSONDecodeError as e:
```

Comment:
Great job catching this error, but I think it is currently unstable. Sometimes it misses the format and fails to convert to json!

I would recommend having a `while count < 10` kind of condition, which allows gemini model to take in the error prone response and error itself, and try to fix it without exceeding the rate limit!

Reply:
Switched to relatively stable llama-3.1 free API calls for now, hence closing this! @liamjdavis